rpcs3/Emu/Cell/Modules/cellVoice.cpp

```diff
@@ -37,177 +147,616 @@ s32 cellVoiceDisconnectIPortFromOPort()
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceGetMuteFlag(portId=%d, bMuted=*0x%x)", portId, bMuted);
+
+	const auto handler = fxm::get<voice_t>();
+	if (!handler)
+	{
+		return CELL_VOICE_ERROR_LIBVOICE_NOT_INIT;
+	}
+
+	if (handler->ports.find(portId) == handler->ports.end())
+	{
+		return CELL_VOICE_ERROR_TOPOLOGY;
+	}
+
+	if (!bMuted)
+	{
+		return CELL_EFAULT;
+	}
+
+	*bMuted = handler->ports[portId].param.bMute;
+
 	return CELL_OK;
 }
 
-s32 cellVoiceGetPortAttr()
+s32 cellVoiceGetPortAttr(u32 portId, s32 attr, vm::ptr<void> attrValue)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceGetPortAttr(portId=%d, attr=*0x%x, attrValue=*0x%x)", portId, attr, attrValue);
+
+	if (!attrValue)
+	{
+		return CELL_VOICE_ERROR_ARGUMENT_INVALID;
+	}
+
+	const auto handler = fxm::get<voice_t>();
+	if (!handler)
+	{
+		return CELL_VOICE_ERROR_LIBVOICE_NOT_INIT;
+	}
+
+	if (handler->ports.find(portId) == handler->ports.end())
+	{
+		return CELL_VOICE_ERROR_TOPOLOGY;
+	}
+
+	if (true)
+	{
+		return CELL_VOICE_ERROR_DEVICE_NOT_PRESENT;
+	}
+
 	return CELL_OK;
 }
 
-s32 cellVoiceGetPortInfo()
+s32 cellVoiceGetPortInfo(u32 portId, vm::ptr<s32> attr)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceGetPortInfo(portId=%d, attr=*0x%x)", portId, attr);
+
+	const auto handler = fxm::get<voice_t>();
+	if (!handler)
+	{
+		return CELL_VOICE_ERROR_LIBVOICE_NOT_INIT;
+	}
+
+	if (handler->ports.find(portId) == handler->ports.end())
+	{
+		return CELL_VOICE_ERROR_TOPOLOGY;
+	}
+
 	return CELL_OK;
 }
 
-s32 cellVoiceGetSignalState()
+s32 cellVoiceGetSignalState(u32 portId, s32 attr, vm::ptr<void> attrValue)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceGetSignalState(portId=%d, attr=%d, attrValue=*0x%x)", portId, attr, attrValue);
+
+	if (!attrValue)
+	{
+		return CELL_VOICE_ERROR_ARGUMENT_INVALID;
+	}
+
+	const auto handler = fxm::get<voice_t>();
+	if (!handler)
+	{
+		return CELL_VOICE_ERROR_LIBVOICE_NOT_INIT;
+	}
+
+	if (handler->ports.find(portId) == handler->ports.end())
+	{
+		return CELL_VOICE_ERROR_TOPOLOGY;
+	}
+
+	if (true)
+	{
+		return CELL_VOICE_ERROR_DEVICE_NOT_PRESENT;
+	}
+
 	return CELL_OK;
 }
 
-s32 cellVoiceGetVolume()
+s32 cellVoiceGetVolume(u32 portId, vm::ptr<f32> volume)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceGetVolume(portId=%d, volume=*0x%x)", portId, volume);
+
+	const auto handler = fxm::get<voice_t>();
+	if (!handler)
+	{
+		return CELL_VOICE_ERROR_LIBVOICE_NOT_INIT;
+	}
+
+	if (handler->ports.find(portId) == handler->ports.end())
+	{
+		return CELL_VOICE_ERROR_TOPOLOGY;
+	}
+
+	if (!volume)
+	{
+		return CELL_EFAULT;
+	}
+
+	*volume = handler->ports[portId].param.volume;
+
 	return CELL_OK;
 }
 
-s32 cellVoiceInit()
+error_code cellVoiceInit(vm::ptr<CellVoiceInitParam> pArg)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	cellVoice.todo("cellVoiceInit(pArg=*0x%x)", pArg);
+
+	if (!pArg)
+	{
+		return CELL_VOICE_ERROR_ARGUMENT_INVALID;
```

Comment:
wrong order as mentioned in other comments, LIBVOICE_INITIALIZED should be the first.